### DISCUSSION TITLE
Loading questions into survey-multi-choice

### DISCUSSION THREAD

**Discussion by emilymech** [jspsych/jsPsych#972] [Q&A] [answered]
Hi everyone,
I'm programming a self-paced reading experiment, and after presenting each of my sentences, I would like to ask a comprehension question unique to the sentence that was just presented. The question and answer options are stored in a csv that is being loaded in (the same method I'm using to present my sentences). However, I'm running into a lot of trouble indexing the question and answer options with the survey-multi-choice plugin.

For reference, I'm loading in my practice trials using the following method, and this is working.

```
  var practice = {
    type: "moving-window",
    words: "spr_sentences",
    choices: key_codes,
    stimulus_duration: max_display_time,
    trial_duration: max_display_time,
    post_trial_gap: iti,
    on_start: function(practice) {
      practice.words = prac_trials[prac_iterator]['spr_sentences'];
      progress_location = "task";
    },
    on_finish: function(data) {
      prac_iterator = prac_iterator + 1
    }
  }

  timeline.push(practice_loop)
```

However, when I try a similar method for my comprehension questions, I am getting an error that the data I'm trying to load in is undefined. 
```
  var comprehension = {    
    type: "survey-multi-choice",
    questions: [
      {prompt: prac_trials[prac_iterator]['comprehension_questions'], options: prac_trials[prac_iterator]['answer_options'], required:true, name: "Comprehension Question"}
    ]
  }
```

My practice loop that includes a function to check if the participant has left the screen, a fixation cross, the practice, and the comprehension questions is pasted below.

```
  //looping through all the practice trials
  var practice_loop = {
    timeline: [ifleft, fixation, ifleft, practice, ifleft, comprehension],
    loop_function: function(data){
      if(prac_iterator < prac_trials.length){
        return true;
      } else {
        return false;
      }
    }
  }
```
Console.log checks of the data (e.g., prac_trials[prac_iterator]['comprehension_questions']) within the practice loop show me that the data is correctly loaded in. I've tried various on_start/on_finish combinations that try to mimic what is working in the practice loop, but to no avail. 

I'm new to jsPsych, so I feel that it is very likely that I'm indexing incorrectly within my comprehension variable, but I haven't been able to figure out the correct method. Does anyone have any suggestions? I'm happy to clarify if any of my post is unclear.

Thank you!!

**Comment by becky-gilbert** [comment, marked as answer]
Hi @emilymech!
The problem is that you're updating the value of `prac_iterator` _while the experiment is running_, but for your comprehension questions, the value of the questions parameter is evaluated _immediately when the page first loads_ (not during the experiment). The reason this works fine for your practice trial is because there you're using the `prac_iterator` value inside of functions that run during the experiment (`on_start` and `on_finish`).
In order to get the current value of `prac_iterator` during the experiment, you can use [dynamic parameters](https://www.jspsych.org/overview/trial/#dynamic-parameters). Specifically, you can try turning your comprehension `questions` parameter into a function. The function will run during the experiment (immediately before the trial executes) and it should return the parameter value that you want for that trial, e.g. 
```javascript 
questions: function() {
 var q_array = [
      {prompt: prac_trials[prac_iterator]['comprehension_questions'], options: prac_trials[prac_iterator]['answer_options'], required:true, name: "Comprehension Question"}
    ];
  return q_array;
}
```

And if you want to check what the value of different variables etc. are when your `questions` function runs, you can just put a `console.log` somewhere in the function before the `return` statement.

**Reply by emilymech** [reply]
@becky-gilbert Thanks for the great explanation and code!! This all makes sense (and runs!) now.
